### ISSUE TITLE
Policy discussion: implicit WASI versus explicit WASI

### ISSUE THREAD

**Issue by joshtriplett** [bytecodealliance/wasmtime#593]
Commit d9edb9521870e086bd7919691beabe719a7dd5d0 made wasmtime-interface-types automatically look for wasm binaries that import WASI, and enable WASI for them (with a default configuration).

For security reasons, a user/embedder of wasmtime may not necessarily want to let a wasm binary use WASI, and especially may not want to allow it if unconfigured. An embedder may instead want to use wasm for pure compute and not give it the full WASI interface, or hand it only a few special-purpose functions, for instance. Or, a user/embedder may have different policies for WASI binaries and more heavily sandboxed wasm binaries with minimal API surface areas.

I'm filing this issue to discuss general policy with respect to enabling WASI. How can we make it relatively convenient to use WASI, while also making it convenient to use wasm without WASI, and not accidentally doing the former when someone expected the latter?

**Comment by sunfishcode** [comment]
One of our goals of WASI is to remove all ambient authorities. The current WASI snapshot does have a few functions with ambient authority, but I expect we'll be fixing those soon. Consequently, one approach is to just leave all the WASI APIs exposed, knowing that without any capability handles, they can't do anything. I expect this won't be enough for all use cases, but it might be a nice default behavior.

**Comment by joshtriplett** [comment]
@sunfishcode That would certainly *help*, but there's still a difference between having the API surface area but no handles and not having the API surface area at all. (Both should absolutely be secure states.) Also, as you observed, WASI isn't that way today.

For example, an embedder may want to have their own WASI backend where "file" means something different and doesn't reference the filesystem, and wouldn't ever want to accidentally invoke a module with the default WASI implementation instead. The problem isn't just that this implicitly enables WASI, it's that it implicitly enables the default WASI implementation even if the user might have intended to supply a different implementation.

For that matter, an embedder may prefer to have a noisy and clear error saying that the module couldn't resolve the WASI functions (similar to trying to run an ELF binary without a required library installed), rather than successfully resolving the functions, running the binary, and then subsequently failing at runtime.

I would personally take the position that we shouldn't autodetect and autoenable WASI unless requested. I like the property of WebAssembly starting with *no* API surface area, such that even the ability to call WASI functions is a kind of authority. By contrast with the idea of a global toggle, that would make it less likely that a missing check somewhere in wasmtime would result in making API available that the caller didn't intend to make available.

But at a minimum, we should have a clear configuration toggle for enabling or disabling the use of WASI, which would in the future appear alongside configuration such as one's own implementation of the WASI functions. And any code in wasmtime that would enable WASI should check that.

Thoughts?

(Also, to clarify something: I *absolutely* think that the wasmtime cli should enable WASI by default; that's the right default for the command-line interface.)

**Comment by peterhuene** [comment]
Link to seemingly related discussion (although talking about the C embedding API): https://github.com/WebAssembly/WASI/issues/149

**Comment by alexcrichton** [comment]
This seems reasonable to me to have as an option, especially because it's a one-liner either way (enable or disable). With the case of the Python extension specifically though the intention is that it integrates wasm modules into Python's module system, so there's not actually any explicit interaction with wasmtime's APIs beyond simply importing it. For handling the Python case I think we'll want to have some ideas about how, when integrating into language module system, wasmtime configuration should be provided. For example such configuration could be done to disable wasi (or enable it), or similarly various compilation options like lightbeam/cranelift/etc.

**Comment by kubkon** [comment]
Like @alexcrichton, I'm also happy with having WASI disabled by default, and have the embedder purposely enable it. FWIW, as @sunfishcode mentioned, there's already some discussion in progress (see [WASI-11-07.md]) on weeding out the ambient authorities.

[WASI-11-07.md]: https://github.com/WebAssembly/WASI/blob/master/meetings/2019/WASI-11-07.md

**Comment by fitzgen** [comment]
I don't think this is applicable anymore today. Wasmtime's embedding API and CLI both have knobs for explicitly enabling/disabling WASI.
